lib/rbUtil.js

```diff
@@ -54,24 +54,65 @@ rbUtil.wildcardSubdomain = function(domain) {
     }
 };
 
+function isHtmlSvgContent(res) {
+    return res.headers
+        && /^(?:text\/html|image\/svg)/i.test(res.headers['content-type']);
+}
+
+var cspHeaderVariants = [
+    'content-security-policy',
+    // For IE 10 & 11
+    'x-content-security-policy',
+    // For Chrome <= v25 (seems to be <1% traffic; should we still care?)
+    'x-webkit-csp',
+];
+
 /**
- * Constructs Content-Security-Policy header to send in response
+ * Adds Content-Security-Policy & related headers to send in response
  *
- * @param domain the domain to allow. If undefined, '*' is allowed
+ * @param res response object
  * @param options options containing the following fields:
- *                - allowInline - if true 'unsafe-inline' is added to style-src
- * @returns {string} CSP header value
+ *                - domain: the domain to allow. If undefined, '*' is allowed
+ *                - allowInline: if true 'unsafe-inline' is added to style-src
  */
-rbUtil.constructCSP = function(domain, options) {
-    var styleSource;
-    if (domain) {
-        styleSource = this.wildcardSubdomain(domain);
-        styleSource = 'http://' + styleSource + ' https://' + styleSource;
+rbUtil.addCSPHeaders = function(res, options) {
+    if (!res.headers) {
+        res.headers = {};
+    }
+    var rh = res.headers;
+    var csp;
+    if (isHtmlSvgContent(res)) {
+        // Our main production clients will ignore CSP anyway (by loading via
+        // XHR or fetch), so we need to sanitize our HTML assuming that no
+        // CSP is enforced on the client. This means that we might actually
+        // gain some security by convincing some users to load HTML in frames,
+        // and thus actually enforce CSP. It is also useful to preview content
+        // in the browser.
+        var styleSource;
+        if (options.domain) {
+            styleSource = this.wildcardSubdomain(options.domain);
+            styleSource = 'http://' + styleSource + ' https://' + styleSource;
+        } else {
+            styleSource = '*';
+        }
+        csp = "default-src 'none'; media-src *; img-src *; style-src "
+            + styleSource
+            + (options && options.allowInline ? " 'unsafe-inline'" : "")
+            + "; frame-ancestors 'self'";
     } else {
-        styleSource = '*';
+        // Other content: Disallow everything, especially framing to avoid
+        // re-dressing attacks.
+        csp = "default-src 'none'; frame-ancestors 'none'";
+    }
+    rh['x-xss-protection'] = '1; mode=block';
+
+    // Let backend services override CSP headers
+    // XXX: Re-consider this policy
+    if (!rh['content-security-policy']) {
```

Comment:
Indeed @Pchelolo, good catch! That was in effect previously, but has been removed in this PR. A slightly more optimised version:

```js
if (rh['content-security-policy']) {
  csp = rh['content-security-policy'];
}
cspHeaderVariants.forEach(function(name) {
  rh[name] = csp;
});
```